File: app/src/main/java/com/adrzdv/mtocp/domain/repository/ViolationRepository.java
package com.adrzdv.mtocp.domain.repository;

import android.database.sqlite.SQLiteConstraintException;

import androidx.room.Query;

import com.adrzdv.mtocp.data.db.entity.ViolationEntity;

import java.util.List;

public interface ViolationRepository {

    List<ViolationEntity> getAll();

    void addNew(ViolationEntity violation);

    void updateByCode(ViolationEntity violation) throws SQLiteConstraintException;

    void saveAll(List<ViolationEntity> list);

}

File: app/src/main/java/com/adrzdv/mtocp/domain/repository/ViolationRepository.kt
package com.adrzdv.mtocp.domain.repository

import com.adrzdv.mtocp.data.db.entity.ViolationDepartmentCrossRef
import com.adrzdv.mtocp.data.db.entity.ViolationDivisionCrossRef
import com.adrzdv.mtocp.data.db.entity.ViolationEntity
import com.adrzdv.mtocp.data.db.entity.ViolationFullInfo
import com.adrzdv.mtocp.data.db.entity.ViolationTypeCrossRef

interface ViolationRepository : BaseAppRepository<ViolationEntity, Int> {
    suspend fun getAllFull(): List<ViolationFullInfo>

    suspend fun saveViolation(
        violation: ViolationEntity,
        departmentIds: List<Int>,
        divisionIds: List<Int>,
        typeIds: List<Int>
    )

    suspend fun syncViolations(
        violations: List<ViolationEntity>,
        departmentLinks: List<ViolationDepartmentCrossRef>,
        divisionLinks: List<ViolationDivisionCrossRef>,
        typeLinks: List<ViolationTypeCrossRef>
    )

    suspend fun clearAllViolationData()

    suspend fun delete(id: Int)
}
